### PR TITLE
feat: Add serializations

### PR DESCRIPTION
**Pull request by boris-marinov**
This `serialize` derivation adds a `toJSON` method to the instance prototype and a static `fromJSON` function to the ADT constructor. 

`fromJSON` does not use the variant constructors but rather it assumes the listed fields are correct and just copies them.

The methods work recursively: `toJSON` looks for `toJSON` method on each value and leaves it as it is if it does not have one.  `fromJSON` acceps as a second argument an object containing ADT definitions that applies them if the types match. 

Looking forward to hearing your thoughts on this.

### CODE REVIEW

File: src/core/adt/serialize.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+const { tagSymbol, typeSymbol } = require('./core');
+const mapValues = require('folktale/core/object/map-values');
+
+const assertType = (given, expected) => {
+  if (expected !== given) {
+    throw new TypeError(`
+       The JSON structure was generated from ${expected}.
+       You are trying to serialize it into ${given}.
```
Comment (robotlolita) [inline src/core/adt/serialize.js]: Should probably be `You are trying to parse it as ${given}`

File: src/core/adt/serialize.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+const assertType = (given, expected) => {
+  if (expected !== given) {
+    throw new TypeError(`
+       The JSON structure was generated from ${expected}.
+       You are trying to serialize it into ${given}. 
+    `);
+  }
+};
+const parseValue = (parsers) => (value) =>
+  parsers[value.typeName] ? parsers[value.typeName].fromJSON(value, parsers)
```
Comment (robotlolita) [inline src/core/adt/serialize.js]: This needs to check for `null` and `undefined`, since JS throws errors when you try to treat those as objects. This could be `value == null ? value : parsers[value.typeName] ? ...`

File: src/core/adt/serialize.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+const serializeValue = (value) =>
+  typeof value.toJSON === 'function' ? value.toJSON() : value;
+
+module.exports = (variant, adt) => {
+  const typeName = adt[typeSymbol];
+  const tagName = variant.prototype[tagSymbol];
+
+  variant.prototype.toJSON = function() {
+    return { typeName, tagName, value: mapValues(this, serializeValue) };
```
Comment (robotlolita) [inline src/core/adt/serialize.js]: Maybe use `@@type` and `@@tag` (or some similar prefix) to make collisions less likely?

**Comment by robotlolita** [comment]
Okay, sorry for taking too long to review this. It looks great :+1: 

I like the idea of passing a parsers object to `.fromJSON`. We might want to provide a simpler way for people to construct these objects for the things they use in their applications, but right now I don't really have anything in mind (tracking these globally is easy, but not modular, and doesn't play too well with Node modules' semantics).

Maybe as a separate discussion, do we want to support [Transit](https://github.com/cognitect/transit-format) as a serialisation format? It's a specification for an extensible serialisation format which can use JSON as an encoding.

**Comment by boris-marinov** [comment]
I am glad you like it.

I added the missing checks (weird that that [jsverify does not test w/ nulls](https://github.com/jsverify/jsverify/issues/170)).

The cool think is that the format is not something fancy. So even this can work:
`type.fromJSON(value, require('folktale/data'))`.

The other think I thought was cool was that by passing the 'contents' of the value you are making an assumption about what types do you expect (that is why I am not automatically adding the root type to the pool).  Related to that, what do you think we should do in case a value of unknown type arrives? Currently we just return it as is.

**Comment by boris-marinov** [comment]
However in order for the current functionality nicely with the built-ins the 'type' string of the ADT must be the same as the name under which it is exported, so:
```const Either = data('folktale:Data.Either', {``` should be ```const Either = data('Either', {```.

Or maybe I should rework it so it uses the type provided...

**Comment by robotlolita** [comment]
We could have a `makeParsers([adt1, adt2, adt3, ...])` function that converts the list of things to a map of them.

**Comment by boris-marinov** [comment]
TL;DR Looks good but cannot imagine how exactly it would look.

Yes I like that idea. The only issue I see is where to put this function. It does not make sense for it to be part of the derivation as people shouldn't need to import the derivation function (or know that it exists at all) in order to use `from/toJSON`. 

The funny thing is that the function looks almost like something generic that could be placed in the `Object` module: 
`fromArray: Array(a) => ( a => String ) => Map(a)`
but not exactly.

Also the fact that you need a helper for the default use case bugs me a little (if we consider the `toJSON`  => `fromJSON` pipeline as the default use case that is).

**Comment by robotlolita** [comment]
True.

I'm assuming that what you meant by "Or maybe I should rework it so it uses the type provided..." was having an object that provides the parsers look like:

```
{
  Either: { [typeName]: String, fromJson: Function },
  Maybe: { ... },
  ...
}
```

Then we can pass `require('folktale/data')` directly, and we can combine parser providers with the `...` object spread, so `const parsers = { ...require('folktale/data'), ...require('my-types') }`.

This is a much simpler API, and uses built-in language features, which is nice.

**Comment by boris-marinov** [comment]
OK, I think this is a good compromise:
The `fromJSON` method rebuilds the dictionary by setting the types of the objects as keys.
This process can be turned off by the `keysIndicateType` flag.
This is also helpful because, after being performed initially, the rebuilding can be turned off during subsequent recursive calls to `fromJSON`.
